pkg/lib/dialogs.tsx

```diff
@@ -153,41 +158,28 @@ export const useDialogs = () => {
 };
 
 export const WithDialogs = ({ children } : { children: React.ReactNode }) => {
-    const is_open = useRef(false); // synchronous
-    const resolveRef = useRef<((x: unknown) => void) | null>(null);
-    const rejectRef = useRef<((x: unknown) => void) | null>(null);
     const [dialog, setDialog] = useState<React.ReactNode>(null);
+    const shown = useRef<"show" | "run" | "close">("close");
+
+    function transition(expected: string, to: typeof shown.current, arg: React.ReactNode = null) {
+        if (shown.current !== expected)
+            throw new Error(`Dialogs.${to}(${JSON.stringify(arg)}) called, but that's only valid` +
+                            `after .${expected}(), current dialog is ${JSON.stringify(dialog)}.`);
+        shown.current = to;
+        setDialog(arg);
+    }
 
     const Dialogs: Dialogs = {
-        show: component => {
-            if (component && is_open.current)
-                console.error("Dialogs.show() called for",
-                              JSON.stringify(component),
-                              "while a dialog is already open:",
-                              JSON.stringify(dialog));
-            is_open.current = !!component;
-            setDialog(component);
-            return new Promise((resolve, reject) => {
-                resolveRef.current = resolve;
-                rejectRef.current = reject;
-            });
-        },
-        close: (args) => {
-            is_open.current = false;
-            setDialog(null);
-            if (resolveRef.current !== null) {
-                resolveRef.current(args);
-                resolveRef.current = null;
-                rejectRef.current = null;
-            }
-        },
-        reject: (err) => {
-            is_open.current = false;
-            setDialog(null);
-            if (rejectRef.current !== null) {
-                rejectRef.current(err);
-                resolveRef.current = null;
-                rejectRef.current = null;
+        show: (component: React.ReactNode) => transition("close", "show", component),
+        close: () => transition("show", "close"),
+        run: async (component, properties) => {
+            try {
+                return await new Promise((resolve, reject) => {
```

Comment:
It might be me misunderstanding promises again, but the function is `async` so returns a `Promise` yet we `await` a new promise here. It looks like we can just return `new Promise` instead. (And drop the `async`).

Reply:
No, you didn't misunderstand. I wrote the same thing (sorry, we reviewed in parallel)

Reply:
The issue is the `finally { }` block.  Also: if I try to write this as `new Promise(...).finally(...);` then I have to annotate it, since the compiler isn't smart enough to infer the promise type from the return type in the presence of `.finally()`.

This issue has me on the edge of filing a bug against TypeScript, but I think I understand it, and I think the compiler is right.  Dodging it via writing an async function seems to be a good approach.

See this [TypeScript playground link](https://www.typescriptlang.org/play/?target=10&ssl=1&ssc=1&pln=2&pc=1#code/PTAEEFQJwUwZwPYBsBuNQDMCuA7AxgC4CWCOopoAhqAA5QIC2Rc6CGVoA5kWmQQJ40YAGlAAVAFAChoAErxkaADxiAfKAC8oABQpKSLDABc4gJSb1KBEQAmAbgkSQETLkIkyAdwAWRPN9AbBHhQRAZ0TwQoAGsqHBtQAm8YMjx9JDhE5OgFVHRsfGJSKUF0eUQ8qBV1LW1YCrQTcsUYavMNS2t7CRgADxoogldCjxyGmAB9AgQJgBY63Ma5RdacLAYAIxgoVXMAbwlQI7GW7VnTBwBfR2dZXA56lqhh9woiHDhbVnZqOkZmGA9fqDF5FVLpCaPSrVBbjKBNFZVNSmEwABXoTBY1VAB2OoGcOAQBGMoAABjgYJ5QOj-ixSYFgnAcAByIYUmAJbGAFAJQEQCMzMngorBCEh+LycBhtrAbIdjrACFgoGQKVSaZiYLCnhcJNcnGA7mRqFC0M8Cq8yO9PjZvhw-hrRJR4qAkAhOFl0EKGDQkDAwfrQABGAB04l8mQYlHFW1AQpwpuJCWmcSocH4+FBHlDAG0ABwABkLADYALpAgZQIbmsGxiEm7YTJ02Cauzgw+vw5ZwtpojEA7G444Eokk8mU6l9ukM+AstkwDmgbm8-mC4UwUXi95SqAyuVHBVKlXj9UArWVUzBjDvdL8bTadrqOMVGDB1vaABEABFSDAAITv0wdT1ZwAGVKHCVMqA2BA0FEDYsCGJJ8jcGtOD9TIkJOPJMwoFAiGoJ1yBoMFMgQDYACt1wIEoZAAeWIuBmmhNRNB0SgoE4OATD2LCliY5QWMuB9QCsWwHD6CsqxQ0YEAYyEVimGZ5h4+tQEuUBuN4kl+NWdYth2NT9j3LSziAxwJJBatRjSJAkAmWSCHk7s1DPU0THoghGMRHsJ1pVoWMHI5h2JEwxzVScYHpIIZ1ZUB2U5FieT5AVYzXDcJW3Xc8QPZU4uPCLXPQDotKgbQVJWQyzIDEMw2YUBI2jT1SATBdk0Iyg0wzKzSBzAtizLCzKxw8FbPs4inKeRt4hbN120RdyGJ0pFdl7PyB2M5warEUpQGZE8sVwaJCU8HBVGZXlMkJIYOs+TgcEoDZfUSBBEh2vaIuqZlgw2sA8W2mRmUO46WQuuKiVTW77se9Bk2kdBmTEL6frxY4EfOoUsCQBIYytAgnWIShE1ATw+QCdqoA2PkoHY8U4eJ3x-FSzHsfQXBYCQQnWpewGcCOhATq+0BswAJgAZmF4WyzxZwAGFvCdbgcHdJI6uTVVfI1GFg218wr16EJMOYOBDG+7K-UPPLwr8wqLBKsqtMqy9r1s2971tp9kBfN8vx-f9AKuG4wFRQZ3mVl6OvTPBQgEJ79EQLSMOyI3DHLEEI666SKBsuyHImyopubVtG06vA5rhBbPKWnz9v89RAsSKBxXr6WwCu0d1ZrqLGVnPKFyXZLVx3KixQy6UOWMs3FVyyhPEoPlLY109VOKjt7dUoSdTxdSr3ul2cQn44Pd9V83Q-b8KT9hwt91CQgA) for more information.

Reply:
Note: I'm also open to change to the `new Promise().finally()` style, but on balance, I don't think it improves readability.

Reply:
Right, I expected `return new Promise().finally() style` to do the exact same without the additional overhead of the async/await pair. If it gets typechecking into trouble, that does feel like a ts limitation/bug (as in the end it should amount to the exact same thing). But not necessary to lose time on this, it just looked like an easy cleanup.